From c949ba5388f242e9b88c9631d7cf8b409bc26331 Mon Sep 17 00:00:00 2001
From: K Prateek Nayak <kprateek.nayak@amd.com>
Date: Fri, 26 Apr 2024 12:46:11 +0200
Subject: sched/eevdf: Limit preemption a little more

TL;DR, adding tasks to the right of the tree moves avg_vruntime right,
which possibly makes curr non-eligible and causes insta-preemption
irrespective of it only having just landed on the CPU.

Signed-off-by: Peter Zijlstra (Intel) <peterz@infradead.org>
---
 kernel/sched/fair.c     |  4 ++--
 kernel/sched/features.h | 11 +++++++----
 2 files changed, 9 insertions(+), 6 deletions(-)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index 8880da0a1900d6..111b828d6c1dd1 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -867,7 +867,7 @@ static inline bool pick_curr(struct cfs_rq *cfs_rq,
 	/*
 	 * Nothing to preserve...
 	 */
-	if (!curr || !sched_feat(RUN_TO_PARITY))
+	if (!curr || !sched_feat(RESPECT_SLICE))
 		return false;
 
 	/*
@@ -875,7 +875,7 @@ static inline bool pick_curr(struct cfs_rq *cfs_rq,
 	 * is consumed. Note: placement of positive lag can push V left and render
 	 * @curr instantly ineligible irrespective the time on-cpu.
 	 */
-	if (!entity_eligible(cfs_rq, curr))
+	if (sched_feat(RUN_TO_PARITY) && !entity_eligible(cfs_rq, curr))
 		return false;
 
 	/*
-- 
cgit 1.2.3-korg

